tqftpserv.c

```diff
@@ -463,16 +477,29 @@ static int handle_reader(struct tftp_client *client)
 	last = buf[2] << 8 | buf[3];
 	// printf("[TQFTP] Got ack for %d\n", last);
 
+	/* We've sent enough data for rsize already */
+	if (last * client->blksize > client->rsize)
+		return 0;
+
 	for (block = last; block < last + client->wsize; block++) {
+		size_t offset = client->seek + block * client->blksize;
+		size_t rsize = 0;
+		/* Check if need to limit response size based for requested rsize */
+		if (block * client->blksize + client->blksize > client->rsize)
+			rsize = client->rsize - (block * client->blksize);
```

Comment:
What about something like this? This way you can drop logic from `tftp_send_data`
```
if ((block + 1) * client->blksize > client->rsize)
    response_size = client->rsize % client->blksize;
else
    response_size = client->blksize;
```

Reply:
Or even better, add `remaining` variable and use it through the code to calculate `response_size`. It can then be used in the 'enough data' check.

Reply:
I don't quite see how can do this, if we don't drop the not-enough-data-read check from tftp_send_data (that check is explained over there https://github.com/linux-msm/tqftpserv/pull/13#discussion_r1544527304)

We need some way to differentiate up-to-x-bytes and exactly-x-bytes which is currently what the 0 in response_size signifies